src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHelpers.cs

```diff
@@ -152,5 +169,274 @@ internal static void AppendData(this IncrementalHash hasher, ReadOnlySpan<byte>
+        {
+            int retVal = 0;
+
+            if (isTagged)
+            {
+                retVal += sizeOfCborTag;
+            }
+
+            retVal += SizeOfArrayOfLessThan24;
+
+            retVal += GetByteStringEncodedSize(encodedProtectedHeadersLength);
+            retVal += CoseHeaderMap.ComputeEncodedSize(unprotectedHeaders);
+
+            if (content is null)
+            {
+                retVal += SizeOfNull;
+            }
+            else
+            {
+                retVal += GetByteStringEncodedSize(content.Length);
+            }
+
+            return retVal;
+        }
+
+        internal static int ComputeSignatureSize(CoseSigner signer)
+        {
+            int keySize = signer.Key.KeySize;
+            KeyType keyType = signer._keyType;
+
+            if (keyType == KeyType.ECDsa)
+            {
+                return 2 * ((keySize + 7) / 8);
+            }
+            else // RSA
+            {
+                Debug.Assert(keyType == KeyType.RSA);
+                return (keySize + 7) / 8;
+            }
+        }
+
+        // If we Validate: The caller did specify a COSE Algorithm, we will make sure it matches the specified key and hash algorithm.
+        // If we Slip: The caller did not specify a COSE Algorithm, we will write the header for them, rather than throw.
+        internal static int? ValidateOrSlipAlgorithmHeader(CoseSigner signer)
+        {
+            int algHeaderValue = GetCoseAlgorithmHeaderFromCoseSigner(signer);
+
+            CoseHeaderMap? protectedHeaders = signer._protectedHeaders;
+            if (protectedHeaders != null && protectedHeaders.TryGetValue(CoseHeaderLabel.Algorithm, out CoseHeaderValue value))
+            {
+                ValidateAlgorithmHeader(value.EncodedValue, algHeaderValue, signer);
+                return null;
+            }
+
+            CoseHeaderMap? unprotectedHeaders = signer._unprotectedHeaders;
+            if (unprotectedHeaders != null && unprotectedHeaders.ContainsKey(CoseHeaderLabel.Algorithm))
+            {
+                throw new CryptographicException(SR.Sign1SignAlgMustBeProtected);
+            }
+
+            return algHeaderValue;
+
+            static void ValidateAlgorithmHeader(ReadOnlyMemory<byte> encodedAlg, int expectedAlg, CoseSigner signer)
+            {
+                int? alg = DecodeCoseAlgorithmHeader(encodedAlg);
+                Debug.Assert(alg.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
+
+                if (expectedAlg != alg.Value)
+                {
+                    KeyType keyType = signer._keyType;
+                    string exMsg;
+                    if (keyType == KeyType.RSA)
+                    {
+                        exMsg = SR.Format(SR.Sign1SignCoseAlgorithDoesNotMatchSpecifiedKeyHashAlgorithmAndPadding, alg.Value, keyType.ToString(), signer.HashAlgorithm.Name, signer.RSASignaturePadding!.ToString());
```

Comment:
This long line could use some chopping, and you shouldn't ever have to call ToString manually...


```suggestion
                        exMsg = SR.Format(
                            SR.Sign1SignCoseAlgorithDoesNotMatchSpecifiedKeyHashAlgorithmAndPadding,
                            alg.Value,
                            keyType,
                            signer.HashAlgorithm.Name,
                            signer.RSASignaturePadding);
```